### PR TITLE
Fix calculation of new block index on drop

### PR DESCRIPTION
**Pull request by oandregal**
Fixes https://github.com/WordPress/gutenberg/issues/9823, https://github.com/WordPress/gutenberg/issues/10202 and https://github.com/WordPress/gutenberg/issues/9303.

### Testing

- Drag any block and drop it at the place you want. Check that it behaves as expected.
- Drag any block and drop it ot its own top/bottom zones ([see](https://github.com/WordPress/gutenberg/issues/10202)).
- Drag a columns block and drop it within one of its columns ([see](https://github.com/WordPress/gutenberg/issues/9303#issuecomment-420563076)).
- Drag a block and drop it in the top zone of any block below it ([see](https://github.com/WordPress/gutenberg/issues/9823)).

### CODE REVIEW

File: packages/editor/src/components/block-draggable/index.js [hunk 1]
```diff
@@ -4,13 +4,12 @@
...
 
-const BlockDraggable = ( { children, clientId, rootClientId, blockElementId, index, layout, onDragStart, onDragEnd } ) => {
+const BlockDraggable = ( { children, clientId, rootClientId, blockElementId, index, onDragStart, onDragEnd } ) => {
 	const transferData = {
 		type: 'block',
-		fromIndex: index,
-		rootClientId,
-		clientId,
-		layout,
+		srcIndex: index,
```
Comment (nosolosw) [inline packages/editor/src/components/block-draggable/index.js]: Improve nomenclature, as it was inconsistent (`fromIndex` but not `fromClientId`?), especially when this data was used in the `BlockDropZone`.

**Review by youknowriad — COMMENTED** [review]
Tested and works properly, I'd appreciate if we can just change the name of `getDescendants`.

Also, it's important to run `npm run docs:build` to regenerate the new selector's docs (and commit the changes).

File: packages/editor/src/store/selectors.js [hunk 2]
```diff
@@ -565,6 +565,20 @@ export const getBlocks = createSelector(
...
+/**
+ * Returns an array containing the clientIds of all descendants
+ * of the blocks given.
+ *
+ * @param {Object} state Global application state.
+ * @param {Array} clientIds Array of blocks to inspect.
+ *
+ * @return {Array} ids of descendants.
+ */
+export const getDescendants = ( state, clientIds ) => flatMap( clientIds, ( clientId ) => {
```
[the remark this answers is not in the document]
Comment (nosolosw) [reply packages/editor/src/store/selectors.js]: Renamed it to `getClientIdsOfDescendants` in bebfdbd.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

File: packages/editor/src/components/block-drop-zone/index.js [hunk 3]
```diff
@@ -56,28 +77,29 @@ class BlockDropZone extends Component {
...
+		const { rootClientId: dstRootClientId, clientId: dstClientId, index: dstIndex, getDescendants } = this.props;
+		const { srcRootClientId, srcClientId, srcIndex, type } = parseDropEvent( event );
+
+		const isBlockDropType = ( dropType ) => dropType === 'block';
+		const isSameLevel = ( srcRoot, dstRoot ) => {
+			// Note that rootClientId of top-level blocks will be undefined OR a void string,
+			// so we also need to account for that case separately.
+			return ( srcRoot === dstRoot ) || ( ! srcRoot === true && ! dstRoot === true );
+		};
+		const isSameBlock = ( src, dst ) => src === dst;
```
[the remark this answers is not in the document]
Comment (nosolosw) [reply packages/editor/src/components/block-drop-zone/index.js]: I know it seems straightforward code, but I like how having the function helps document the intent for the reader.

**Pull request merged**